# Import Libraries & Files

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# establish path to project folder
path = r'C:\Users\BrianDavis\OneDrive - abetterchicago.org\Documents\Data Analysis\Achievement 5 (Python)\04-2023 Instacart Basket Analysis'

In [3]:
# import "orders" file
df_ords = pd.read_csv(os.path.join(path,'02 Data','Original Data','orders.csv'), index_col = False)

In [4]:
# import "products" file
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'), index_col = False)

In [5]:
# import "departments" file
df_deps = pd.read_csv(os.path.join(path,'02 Data','Original Data','departments.csv'), index_col = False)

# Exercise 4.4 Wrangling Procedures

In [6]:
#remove "eval_set" column from orders file
df_ords = df_ords.drop(columns = ['eval_set'])

In [7]:
#check eval_set was removed from orders file
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [8]:
#count number of missing values in orders file (n=206209)
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [9]:
#rename "order_dow" column orders file
df_ords.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [10]:
#change data type of order_id column to string
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [11]:
#check to ensure order_id doesn't show up in describe function
df_ords.describe()

user_id  order_number  order_day_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06       3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01       2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01       2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00       0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00       1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01       3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01       5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02       6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [12]:
#transpose departments file
df_deps = df_deps.T

In [13]:
#check to ensure departments file is transposed
df_deps.head()

0
department_id  department
1                  frozen
2                   other
3                  bakery
4                 produce

In [14]:
#add index column to departments file (note following steps undertaken to replace faulty header information in the departnments file
df_deps.reset_index() 

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [15]:
#establish new_header variable as information contained in first row of departments file
new_header = df_deps.iloc[0]

In [16]:
#create new departments file that excludes the first row of data
df_deps_new = df_deps[1:]

In [17]:
#rename columns of new departments file to new_header variable
df_deps_new.columns = new_header

In [18]:
#check to ensure new departments file is formatted correctly
df_deps_new.head()

department_id department
1                 frozen
2                  other
3                 bakery
4                produce
5                alcohol

In [19]:
#establish data_dict variable as the data dictionary for the new departments file
data_dict = df_deps_new.to_dict('index')

In [20]:
#to illustrate data dictionary capabilities 
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [21]:
#to illustrate data dictionary capabilities (continued)
print(data_dict.get('19'))

{'department': 'snacks'}


In [22]:
#establish snacks dataframe, which is a subset of the products file filtered for the department id of 19 (snacks)
df_snacks = df_prods[df_prods['department_id']==19]

In [23]:
#similar to line above, using .loc to pull the location 
df_snacks2 = df_prods.loc[df_prods['department_id']==19]

In [24]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

# Assignment 4.4 

### step 2 - Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [25]:
#per table below, "user_id" column should be considered a string
df_ords.describe()

user_id  order_number  order_day_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06       3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01       2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01       2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00       0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00       1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01       3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01       5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02       6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [26]:
#change "user_id" column to string
df_ords['user_id'] = df_ords['user_id'].astype('str')

### step 3 - Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.

In [27]:
df_ords.rename(columns = {'order_number' : 'order_number_by_user'}, inplace = False)

order_id user_id  order_number_by_user  order_day_of_week  \
0        2539329       1                     1                  2   
1        2398795       1                     2                  3   
2         473747       1                     3                  3   
3        2254736       1                     4                  4   
4         431534       1                     5                  4   
...          ...     ...                   ...                ...   
3421078  2266710  206209                    10                  5   
3421079  1854736  206209                    11                  4   
3421080   626363  206209                    12                  1   
3421081  2977660  206209                    13                  1   
3421082   272231  206209                    14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

### step 4 - Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.


In [28]:
#count values of orders file 
df_ords.value_counts('order_hour_of_day')

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
dtype: int64

#### Observations: 10am is the busiest hour for placing orders, followed by 11am and 3pm; 3am is the least busy


### step 5 - Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [39]:
#print data dictionary for the value of 4
print(data_dict.get('4'))

{'department': 'produce'}


#### Observation: a value of 4 in the "department_id" column refers to alcohol


### step 6 - The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [31]:
#breakfast items are department_id 14
df_breakfast = df_prods.loc[df_prods['department_id'] == 14]

### step 7 - They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [32]:
#dinner parties dataframe will include alcohol (department_id = 5), deli (20), beverages (7), and meat/seafood (12)
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

### step 8 It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [41]:
#call information on dinner parties dataframe
df_dinner_parties.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7650 entries, 2 to 49688
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     7650 non-null   int64  
 1   product_name   7647 non-null   object 
 2   aisle_id       7650 non-null   int64  
 3   department_id  7650 non-null   int64  
 4   prices         7650 non-null   float64
dtypes: float64(1), int64(3), object(1)
memory usage: 358.6+ KB


#### Observation: the dinner parties dataframe has 7,650 rows and no missing values

### step 9 - Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [34]:
#create subset with filter for "user_id" = 1
df_ords_customer_one = df_ords.loc[df_ords['user_id'] == '1']

In [35]:
#call all available information on customer one
df_ords_customer_one

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   
5   3367565       1             6                  2                  7   
6    550135       1             7                  1                  9   
7   3108588       1             8                  1                 14   
8   2295261       1             9                  1                 16   
9   2550362       1            10                  4                  8   
10  1187899       1            11                  4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

#### Observation: there is one missing value for customer one (days since prior order for their first order), but this is a logical missing value since it's their first order. No other missing values are present. 

### step 10 - You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [36]:
df_ords_customer_one.describe()

order_number  order_day_of_week  order_hour_of_day  \
count     11.000000          11.000000          11.000000   
mean       6.000000           2.636364          10.090909   
std        3.316625           1.286291           3.477198   
min        1.000000           1.000000           7.000000   
25%        3.500000           1.500000           7.500000   
50%        6.000000           3.000000           8.000000   
75%        8.500000           4.000000          13.000000   
max       11.000000           4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

#### Observations: customer one placed all of their orders between the first (Saturday) and fourth (Tuesday) day of the week; their average hour of the day in which orders were placed was 10am; they placed orders on average 19 days apart, with a minimum of 0 (ie same day) and a maximum of 30 days  

# Exporting Dataframes

In [37]:
#export orders file 
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [38]:
#export departments file 
df_deps_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))